Kernel/Net: Cleanup the ICMP handler

sdomi opened the pull request:
this may be a bit opinionated, but I think it improves readability by quite a lot + the cast is safer :)

---

This change fixes up handle_icmp to use a bit_cast instead of a reinterpret_cast, which got used due to legacy reasons.

A few dbgln's got turned into dbgln_if's, in accordance with the behavior throughout the rest of the file.

Furthermore, we revert a part of commit ad73adef5d2f8 which introduced inconsistency between struct names. There's no ICMPv4, only ICMP and ICMPv6. Thus, it makes more sense to revert the ICMPv4Type back to ICMPType.

We also remove the FIXME about the TTL - as per RFC1700, TTL of 64 is a recommended default, so the current behavior can stay.

nico commented:
This bit_casts a pointer type if I read this correctly. How is that useful? I thought bit_cast() is for reinterpreting the bits of a literal (think int to float, or things like that).

Looking around a bit, eg https://stackoverflow.com/questions/67765898/aliasing-accesses-through-a-stdbit-casted-pointer seems to back this up.

nico commented on Kernel/Net/NetworkTask.cpp:
```diff
@@ -224,6 +224,8 @@ void handle_ipv4(EthernetFrameHeader const& eth, size_t frame_size, UnixDateTime
 {
     auto& icmp_header = *static_cast<ICMPHeader const*>(ipv4_packet.payload());
+    size_t icmp_packet_size = ipv4_packet.payload_size();
+    size_t icmp_header_size = sizeof(ICMPEchoPacket);
```
nit: maybe make const (both)

nico reviewed (commented):
Commit desc: "cleanup" is a noun, "clean up" the verb

Otherwise seems fine.

For dbgln()s, I personally use the _if variant of it's only interesting to people working in that area and the unconditional form of it might be interesting to others (say, for "not yet implemented" and things like that). But I don't have a strong opinion on that.

It's really only the bit_cast change that feels Feng shui. The rest is fine.

sdomi commented:
> This bit_casts a pointer type if I read this correctly.

my decision to change that `reinterpret_cast` into `bit_cast` mostly stemmed from me reading about C++ cast types the other day and coming into the conclusion that "reinterpret_cast bad". but yes, you're not wrong that in this place, the outcome will likely always be the same.

Should I revert this change back?

kleinesfilmroellchen commented:
> This bit_casts a pointer type if I read this correctly. How is that useful? I thought bit_cast() is for reinterpreting the bits of a literal (think int to float, or things like that).

bit_cast is not just a reinterpreting of bits, that’s what reinterpret_cast is. bit_cast has “as-if” equivalence to memcpy (which makes its behavior, especially for padding, very much defined) and is usable at compile time. In this case this of course works out to identical behavior to reinterpret_cast.

The main point in this case is to guarantee that we’re actually handling pointers since bit_cast checks sizes.

As a rule (not yet fixed in documentation, but this is from what I have talked about with others, especially in the kernel space), we don’t want to use reinterpret_casts, since all of their use cases that don’t immediately invoke UB are covered by other casts, especially bit_cast. (I have been very vocal about this in reviews, and have not received significant pushback.) There’s a lot of legacy reinterpret_casts in the kernel, and as we’re moving through modernizing parts of it (like here), we can change them to the safer alternatives.

nico commented:
> bit_cast is not just a reinterpreting of bits, that’s what reinterpret_cast is.

While you can use reinterpret_cast for reinterpreting bits, it's UB. reinterpret_cast can only be used to cast a pointer to void* and back again (…assuming you want to deref the pointer). The portable way to do bit reinterpretation was to memcpy a value of one type into a value of a different type. That function was usually called bit_cast. When C++11 arrived, the problem with that was that it wasn't constexpr, so bit_cast() became a built-in that has the semantics of that memcpy, while being constexpr. (…as you write.)

This means bit_cast is great for replacing `*reinterpret_cast<T*>(foo>)` (note the initial `*`) for casting valus.

reinterpret_cast does look dangerous because it is. Replacing reinterpret_cast with bit_cast for not-immediately-dereferneced pointer casts does nothing for you, and is not any more modern. It just makes the code look like the author doesn't know what bit_cast does. (See link in https://github.com/SerenityOS/serenity/pull/25000#issuecomment-2336824663). So yes, "reinterpret_cast bad", but this use of bit_cast is just as bad – it has the same semantics.

ADKaster commented:
The sort of "Common Initial Sequence" wrangling we're doing here is not a proper use of bit_cast. 

If we're cleaning up this header and class there's a few things that are strange to me:

1. why is ICMPType not an enum class?
2. why is ICMPHeader not declared using a flexible array member?

```
struct ICMPHeader {
    u8 m_type { 0 };
    u8 m_code { 0 };
    NetworkOrdered<u16> m_checksum { 0 };
    u32 _packet_defined_header_bytes { 0 };
    u8 payload[];
};
```

3. Why do we check the size *after* doing our cast + dereference? That feels ripe for a compiler to delete our check because it would have been UB to access the object if the size were not suitable.

sdomi commented:
@ADKaster 
Ad 1. and 2. - great points! fixed

Ad 3. - ICMP Echo packets can have no payload. That is, `ping -s 0 127.1` will send just the header; This check is to make sure that we actually have anything to copy, *not* whether the ICMP packet is suitable for getting handled. Now, we could just call the memcpy with a 0 length and not care, as far as I'm aware that doesn't introduce any problems. Should we call memcpy instead?

As for the suggestion from your snippet to remove the `[[gnu::packed]]` (unless that was just pseudocode and I read way too much into it), this struct needs to be aligned (as previously outlined in #25009):

```
30.111 [#0 Network Task(9:9)]: handle_arp: Responding to ARP request for my IPv4 address (10.254.254.1)
[Network Task(9:9)]: KUBSAN: reference binding to misaligned address 0x000000200d909022 of type 'const ICMPHeader'
[Network Task(9:9)]: KUBSAN: at /home/domi/serenity/Kernel/Net/NetworkTask.cpp, line 226, column: 25
[Network Task(9:9)]: Kernel + 0x0000000000434064  print_location(AK::UBSanitizer::SourceLocation const&) +0xfc
[Network Task(9:9)]: Kernel + 0x0000000000435564  __ubsan_handle_type_mismatch_v1 +0x31c
[Network Task(9:9)]: Kernel + 0x00000000003ea2b2  Kernel::NetworkTask_main(void*) +0x224a
[Network Task(9:9)]: UB is configured to be deadly, halting the system.
```

@nico your explanation makes sense, I turned it back into reinterpret_cast. I also fixed one other C-style cast just above that all of us have somehow missed (oops).
